File: 백준/Silver/5397. 키로거/키로거.py
from collections import deque
import sys

input = sys.stdin.readline
print = sys.stdout.write

# 테스트 케이스 수 입력
T = int(input().rstrip())

# 각 테스트 케이스 처리
for _ in range(T):
    # 키 입력 문자열
    keys = input().rstrip()
    
    # 커서 기준으로 왼쪽과 오른쪽을 deque로 관리
    left = deque()
    right = deque()
    
    # 키 입력 처리
    for key in keys:
        if key == '-':  # 백스페이스: 커서 앞 문자 삭제
            if left:
                left.pop()
        elif key == '<':  # 왼쪽 이동: 커서 왼쪽으로
            if left:
                right.appendleft(left.pop())
        elif key == '>':  # 오른쪽 이동: 커서 오른쪽으로
            if right:
                left.append(right.popleft())
        else:  # 일반 문자: 커서 위치에 추가
            left.append(key)
    
    # 결과 출력 (left + right 합침)
    print(''.join(left) + ''.join(right) + '\n')

File: 백준/Silver/5397. 키로거/키로거.java
import java.io.*;
import java.util.*;

public class Main {
    static String stkToString(Deque<Character> stk1, Deque<Character> stk2){
        var sb1 = new StringBuilder();
        while(!stk1.isEmpty()) sb1.append(stk1.pop());
        var sb2 = new StringBuilder();
        while(!stk2.isEmpty()) sb2.append(stk2.pop());
        return sb1.reverse().append(sb2).toString();
    }
    public static void main(String[] args) throws IOException {
        var in = new BufferedReader(new InputStreamReader(System.in));
        var out = new PrintWriter(System.out);

        int n = Integer.parseInt(in.readLine());

        for(int i = 0; i < n; i++){
            var left = new ArrayDeque<Character>();
            var right  = new ArrayDeque<Character>();
            String s = in.readLine();
            for(char c : s.toCharArray()){
                if(c == '-') {
                    if(!left.isEmpty()) left.pop();
                } else if(c == '<'){
                    if(!left.isEmpty()) right.push(left.pop());
                } else if(c == '>'){
                    if(!right.isEmpty()) left.push(right.pop());
                } else {
                    left.push(c);
                }
            }
            out.println(stkToString(left, right));
        }



        out.flush();
        out.close();
    }
}